### PR TITLE
CORE-2433 - String Resource not recognized by IntegrationValidator

### PR DESCRIPTION
**Pull request by agerstner-branch**
Refactor the way ApkParser encodes resource ids and fix BranchIntegrationModel to property decode resource strings for scheme and applinks.

## Reference
CORE-2433 - [AndroidManifest] String Resource not recognized by IntegrationValidator

## Description
GitHub Issue #915

When using a string resource to define the custom scheme in the AndroidManifest.xml file (e.g. @string/custom_scheme). The IntegrationValidator doesn't recognize the setup correctly, as it doesn't resolve the actual associated resource, it only tries to compare the expected path with the id.

This caused quite some confusion on my end, as the custom scheme was validated to be working, yet branch refused to pass this integration test. Supporting string resources is especially useful when managing multiple app flavors.

## Testing Instructions
Repro steps:

1. Invoke IntegrationValidator.validate(this) within launcher activity (it should pass through 10 steps)
2. Change AndroidManifest.xml, move intent scheme to string resource: 
```
<!-- Branch URI Scheme -->
<intent-filter>
    <data
        android:host="open"
        android:scheme="@string/custom_scheme" />
    <action android:name="android.intent.action.VIEW" />

    <category android:name="android.intent.category.DEFAULT" />
    <category android:name="android.intent.category.BROWSABLE" />
</intent-filter>
```
3. Repeat step 1, note the following. (It should continue to pass through 10 steps.)
```
... snip ...
2021-09-30 14:03:28.893 17026-17026/com.example.helloworld D/BranchSDK_Doctor: 3. Verifying application package name ... 
2021-09-30 14:03:28.893 17026-17026/com.example.helloworld D/BranchSDK_Doctor: Passed
2021-09-30 14:03:28.893 17026-17026/com.example.helloworld D/BranchSDK_Doctor: 4. Checking Android Manifest for URI based deep link config ... 
2021-09-30 14:03:28.894 17026-17026/com.example.helloworld D/BranchSDK_Doctor: Passed
2021-09-30 14:03:28.894 17026-17026/com.example.helloworld D/BranchSDK_Doctor: 5. Verifying URI based deep link config with Branch dash board. ... 
2021-09-30 14:03:28.894 17026-17026/com.example.helloworld D/BranchSDK_Doctor: Passed
2021-09-30 14:03:28.894 17026-17026/com.example.helloworld D/BranchSDK_Doctor: 6. Verifying intent for receiving URI scheme. ... 
2021-09-30 14:03:28.895 17026-17026/com.example.helloworld D/BranchSDK_Doctor: Passed
2021-09-30 14:03:28.895 17026-17026/com.example.helloworld D/BranchSDK_Doctor: 7. Checking AndroidManifest for AppLink config. ... 
2021-09-30 14:03:28.895 17026-17026/com.example.helloworld D/BranchSDK_Doctor: Passed
2021-09-30 14:03:28.895 17026-17026/com.example.helloworld D/BranchSDK_Doctor: 8. Verifying any supported custom link domains. ... 
2021-09-30 14:03:28.895 17026-17026/com.example.helloworld D/BranchSDK_Doctor: Passed
2021-09-30 14:03:28.896 17026-17026/com.example.helloworld D/BranchSDK_Doctor: 9. Verifying default link domains integrations. ... 
2021-09-30 14:03:28.896 17026-17026/com.example.helloworld D/BranchSDK_Doctor: Passed
2021-09-30 14:03:28.896 17026-17026/com.example.helloworld D/BranchSDK_Doctor: 10. Verifying alternate link domains integrations. ... 
2021-09-30 14:03:28.896 17026-17026/com.example.helloworld D/BranchSDK_Doctor: Passed
2021-09-30 14:03:28.896 17026-17026/com.example.helloworld D/BranchSDK_Doctor: Passed
2021-09-30 14:03:28.896 17026-17026/com.example.helloworld D/BranchSDK_Doctor: --------------------------------------------
    Successfully completed Branch integration validation. Everything looks good!
```

## Risk Assessment [`HIGH` || `MEDIUM` || `LOW`]

- [x] I, the PR creator, have tested — integration, unit, or otherwise — this code.

## Reviewer Checklist (To be checked off by the reviewer only)

- [ ] JIRA Ticket is referenced in PR title.
- Correctness & Style
    - [ ] Conforms to [AOSP Style Guides](https://source.android.com/setup/contribute/code-style)
    - [ ] Mission critical pieces are documented in code and out of code as needed.
- [ ] Unit Tests reviewed and test issue sufficiently.
- [ ] Functionality was reviewed in QA independently by another engineer on the team.

### CODE REVIEW

**Review by echo-branch — APPROVED** [review]

File: Branch-SDK/src/test/java/io/branch/referral/BranchUtilTest.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    @Test
+    public void testEncodeResourceId() {
+        for (int i = 0; i < resourceIds.length; i++) {
+            Assert.assertEquals(encodedIds[i], BranchUtil.encodeResourceId(resourceIds[i]));
+        }
+    }
+
+    @Test
+    public void testDecodeResourceId() {
+        Assert.assertEquals(resourceIds[6], BranchUtil.decodeResourceId(encodedIds[6]));
```
Comment (bklastaitis-branch) [inline Branch-SDK/src/test/java/io/branch/referral/BranchUtilTest.java:65]: ?

File: Branch-SDK/src/test/java/io/branch/referral/BranchUtilTest.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            }
+        }
+
+        // Run all negative tests, verify that the original json document is unchanged
+        for (String replaceNegativeJsonTest : replaceNegativeJsonTests) {
+            try {
+                JSONObject jsonObject1 = new JSONObject(replaceNegativeJsonTest);
+                JSONObject jsonObject2 = new JSONObject(replaceNegativeJsonTest);
+                BranchUtil.replaceJsonKey(jsonObject2, "field", "replaced");
+                Assert.assertEquals(jsonObject1.toString(), jsonObject2.toString());
```
Comment (bklastaitis-branch) [inline Branch-SDK/src/test/java/io/branch/referral/BranchUtilTest.java:95]: Should prolly use a jsonassert library for this comparison

File: Branch-SDK/src/test/java/io/branch/referral/BranchUtilTest.java (hunk 1 above)
Comment (agerstner-branch) [reply Branch-SDK/src/test/java/io/branch/referral/BranchUtilTest.java:65]: good catch, will remove. (was debugging that test case).

File: Branch-SDK/src/main/java/io/branch/referral/BranchUtil.java [hunk 3]
```diff
@@ -158,6 +160,56 @@ static JSONObject addSource(JSONObject params) {
...
+     * @param replaceKey    the new name for the element
+     */
+    public static void replaceJsonKey(JSONObject jsonObject, String findKey, String replaceKey) {
+        Object findObj = jsonObject.opt(findKey);
+        Object replaceObj = jsonObject.opt(replaceKey);
+        if(findObj != null && replaceObj == null && !findKey.equals(replaceKey)) {
+            try {
+                jsonObject.putOpt(replaceKey, findObj);
+            } catch (JSONException e) {
+                // If there was a problem, don't do the remove.
```
[the remark this answers is not in the document]
Comment (agerstner-branch) [reply Branch-SDK/src/main/java/io/branch/referral/BranchUtil.java:206]: Adding a null check before

**Pull request merged**